lib/AST/RequirementMachine/RequirementLowering.cpp

```diff
@@ -817,44 +800,59 @@ void swift::rewriting::expandDefaultRequirements(ASTContext &ctx,
     //       func f() where Self: ~Copyable
     //     }
     //
-    if (defaults.count(subject) == 0) {
+    if (notInScope) {
       errors.push_back(
-          RequirementError::forInvalidInverseOuterSubject(req, structReq.loc));
-      return true;
+          RequirementError::forInvalidInverseOuterSubject(inverse));
+      continue;
     }
 
-    auto currentDefaults = defaults[subject];
-    auto inverseKind = inverse->getInverseKind();
+    auto state = inverses.getOrInsertDefault(canSubject);
 
-    // Check if this inverse is redundant.
-    if (!currentDefaults.contains(inverseKind)) {
+    // Check if this inverse has already been seen.
+    auto inverseKind = inverse.getKind();
+    if (state.contains(inverseKind)) {
       errors.push_back(
-          RequirementError::forRedundantRequirement(req, structReq.loc));
-      return true;
+          RequirementError::forRedundantInverseRequirement(inverse));
+      continue;
     }
 
-    // Apply the inverse to the subject's defaults.
-    currentDefaults.remove(inverseKind);
-    defaults[subject] = currentDefaults;
+    state.insert(inverseKind);
+    inverses[canSubject] = state;
+  }
 
-    // Remove this inverse conformance requirement.
-    return true;
-  }), result.end());
+  // Scan the structural requirements and cancel out any inferred requirements
+  // based on the inverses we saw.
+  result.erase(llvm::remove_if(result, [&](StructuralRequirement structReq) {
+    auto req = structReq.req;
 
-  for (auto &[subject, remainingDefaults] : defaults) {
-    // Expand the remaining default conformance requirements for this subject.
-    for (auto ip : remainingDefaults) {
-      ProtocolDecl *decl = ctx.getProtocol(getKnownProtocolKind(ip));
-      if (!decl) llvm_unreachable("missing known protocol!");
+    if (req.getKind() != RequirementKind::Conformance)
+      return false;
 
-      SourceLoc loc = decl->getLoc();
-      Type protocolType = decl->getDeclaredInterfaceType();
+    // Only consider inferred conformances...
+    if (!structReq.inferred)
+      return false;
 
-      // Add the protocol as an "inferred" requirement.
-      Requirement req(RequirementKind::Conformance, subject, protocolType);
-      result.push_back({req, loc, /*wasInferred=*/true});
-    }
-  }
+    // to an invertible protocol.
+    llvm::Optional<InvertibleProtocolKind> proto;
+    if (auto constraint = req.getSecondType()->getAs<ProtocolType>())
```

Comment:
Desugared requirements always have a ProtocolType on the RHS and you can call `req.getProtocolDecl()` to get the ProtocolDecl